data/src/main/java/com/linkedin/data/schema/compatibility/AnnotationCompatibilityChecker.java

```diff
@@ -0,0 +1,193 @@
+   * @return List<AnnotationCompatibilityResult>
+   */
+  static List<AnnotationCompatibilityResult> getCompatibilityResult(Map<PathSpec, Pair<SchemaAnnotationHandler.AnnotationCheckContext,
+      Map<String, Object>>> olderResolvedPropertiesMap, Map<PathSpec, Pair<SchemaAnnotationHandler.AnnotationCheckContext,
+      Map<String, Object>>> newerResolvedPropertiesMap, List<SchemaAnnotationHandler> handlers)
+  {
+    List<AnnotationCompatibilityResult> results = new ArrayList<>();
+    for (SchemaAnnotationHandler handler : handlers)
+    {
+      String annotationNamespace = handler.getAnnotationNamespace();
+      for (Map.Entry<PathSpec, Pair<SchemaAnnotationHandler.AnnotationCheckContext, Map<String, Object>>> e : olderResolvedPropertiesMap
+          .entrySet())
+      {
+        PathSpec pathSpec = e.getKey();
+        Pair<SchemaAnnotationHandler.AnnotationCheckContext, Map<String, Object>> olderCheckContextAndResolvedProperty =
+            olderResolvedPropertiesMap.get(pathSpec);
+        Map<String, Object> olderResolvedProperties = olderCheckContextAndResolvedProperty.getValue();
+
+        if (newerResolvedPropertiesMap.containsKey(pathSpec))
+        {
+          // If older schema node and newer schema node have the same pathSpec,
+          // they may or may not contain the same annotation namespace as SchemaAnnotationHandler, we need to further check.
+          Pair<SchemaAnnotationHandler.AnnotationCheckContext, Map<String, Object>>
+              newerCheckContextAndResolvedProperty = newerResolvedPropertiesMap.get(pathSpec);
+          Map<String, Object> newerResolvedProperties = newerCheckContextAndResolvedProperty.getValue();
+
+          // If olderResolvedProperties or newerResolvedProperties contains the same namespace as the provided SchemaAnnotationHandler,
+          // we do the annotation check
+          if (olderResolvedProperties.containsKey(annotationNamespace) || newerResolvedProperties.containsKey(
+              annotationNamespace))
+          {
+            AnnotationCompatibilityResult result =
+                handler.annotationCompatibilityCheck(olderResolvedProperties, newerResolvedProperties,
+                    olderCheckContextAndResolvedProperty.getKey(), newerCheckContextAndResolvedProperty.getKey());
+            if (!result.isCheckAnnotation())
+            {
+              _log.info("There is no implementation of annotationCompatibilityCheck in SchemaAnnotationHandler: "
+                  + handler.getClass().getName());
+              break;
+            }
+            results.add(result);
+          }
+          newerResolvedPropertiesMap.remove(pathSpec);
+        }
+        else
+        {
+          if (olderResolvedProperties.containsKey(annotationNamespace))
+          {
+            // olderResolvedPropertiesMap has a PathSpec which the newResolvedPropertiesMap does not have,
+            // it could be either newer schema removed annotation or the field is not existed, pass
```

Comment:
or the field was deleted.

The logic here would be simple if you returned the empty resolveProperties instead if removing it (in AnnotationCheckResolvedPropertiesVisitor) 
So the only case to handle is deleted field. At this point, does it make sense to do annotation compatibility checking? Would the schema compatibility checking be enough?

Reply:
Yes, I agree. 
I will return empty resolvedProperties in AnnotationCheckResolvedPropertiesVistor.
Then, we do not need this else branch check at all.

Reply:
I think we should provide such check to the annotation check, if an existing field is removed. Users should decide whether they will do something or not.